File: src/services/tunnel.ts
import { invoke } from '@tauri-apps/api/core';

export interface TunnelConfig {
  main_domain: string;
  subdomain: string;
  target_port: number;
  health_check_path: string;
  fallback_urls: string[];
}

export interface TunnelStatus {
  is_active: boolean;
  current_mode: 'MainDomain' | 'Fallback' | 'Direct';
  main_domain_status: boolean;
  fallback_status: boolean;
  last_check: number;
  response_time_ms: number;
}

export enum TunnelMode {
  MainDomain = 'MainDomain',
  Fallback = 'Fallback',
  Direct = 'Direct'
}

class TunnelService {
  private config: TunnelConfig | null = null;
  private status: TunnelStatus | null = null;
  private healthCheckInterval: NodeJS.Timeout | null = null;
  private readonly HEALTH_CHECK_INTERVAL = 30000; // 30 seconds

  async initialize(): Promise<void> {
    try {
      // Get initial config and status
      this.config = await this.getConfig();
      this.status = await this.getStatus();
      
      // Start health monitoring
      this.startHealthMonitoring();
      
      console.log('Tunnel service initialized', { config: this.config, status: this.status });
    } catch (error) {
      console.error('Failed to initialize tunnel service:', error);
      
      // Fallback config for single domain
      this.config = {
        main_domain: 'https://www.zentaira.com', // www yönlendirmesi
        subdomain: 'whisper-weather',
        target_port: 3000,
        health_check_path: '/', // Ana sayfayı kontrol et
        fallback_urls: ['https://zentaira.com']
      };
    }
  }

  async getConfig(): Promise<TunnelConfig> {
    try {
      return await invoke<TunnelConfig>('get_tunnel_config');
    } catch (error) {
      console.error('Failed to get tunnel config:', error);
      throw error;
    }
  }

  async updateConfig(config: TunnelConfig): Promise<void> {
    try {
      await invoke('update_tunnel_config', { config });
      this.config = config;
      console.log('Tunnel config updated', config);
    } catch (error) {
      console.error('Failed to update tunnel config:', error);
      throw error;
    }
  }

  async getStatus(): Promise<TunnelStatus> {
    try {
      const status = await invoke<TunnelStatus>('get_tunnel_status');
      this.status = status;
      return status;
    } catch (error) {
      console.error('Failed to get tunnel status:', error);
      throw error;
    }
  }

  async checkHealth(): Promise<TunnelStatus> {
    try {
      const status = await invoke<TunnelStatus>('check_tunnel_health');
      this.status = status;
      return status;
    } catch (error) {
      console.error('Failed to check tunnel health:', error);
      throw error;
    }
  }

  async switchMode(mode: TunnelMode): Promise<void> {
    try {
      await invoke('switch_tunnel_mode', { mode });
      console.log('Switched to tunnel mode:', mode);
      
      // Refresh status after mode switch
      await this.getStatus();
    } catch (error) {
      console.error('Failed to switch tunnel mode:', error);
      throw error;
    }
  }

  private startHealthMonitoring(): void {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
    }

    this.healthCheckInterval = setInterval(async () => {
      try {
        await this.checkHealth();
      } catch (error) {
        console.error('Health check failed:', error);
      }
    }, this.HEALTH_CHECK_INTERVAL);
  }

  // Get the best available URL for general API requests
  getApiUrl(): string {
    // Only use tunnel for health checks, not for API requests
    // API requests should go directly to the backend
    const port = this.config?.target_port || 1420;
    
    // In production Tauri, use tauri://localhost for API calls
    if (typeof window !== 'undefined' && window.__TAURI__) {
      return 'tauri://localhost/api';
    }
    
    return `http://localhost:${port}/api`;
  }

  // Get the weather API URL (direct to weather service)
  getWeatherApiUrl(): string {
    // Weather API should go directly to weather service, not through Zentaira
    const port = this.config?.target_port || 1420;
    
    // In production Tauri, use tauri://localhost for API calls
    if (typeof window !== 'undefined' && window.__TAURI__) {
      return 'tauri://localhost/api/weather';
    }
    
    return `http://localhost:${port}/api/weather`;
  }

  // Get the auth API URL (direct to auth service)
  getAuthApiUrl(): string {
    // Auth API should go directly to auth service, not through Zentaira
    const port = this.config?.target_port || 1420;
    
    // In production Tauri, use tauri://localhost for API calls
    if (typeof window !== 'undefined' && window.__TAURI__) {
      return 'tauri://localhost/api/auth';
    }
    
    return `http://localhost:${port}/api/auth`;
  }

  // Make API requests with automatic fallback
  async makeRequest(url: string, options: RequestInit = {}): Promise<Response> {
    // For API requests, always use the appropriate backend URL
    // Tunnel is only for health checks and routing status
    const port = this.config?.target_port || 1420;
    
    // In production Tauri, use tauri://localhost for API calls
    let fullUrl: string;
    if (typeof window !== 'undefined' && window.__TAURI__) {
      fullUrl = url.startsWith('http') || url.startsWith('tauri://') ? url : `tauri://localhost${url}`;
    } else {
      fullUrl = url.startsWith('http') ? url : `http://localhost:${port}${url}`;
    }

    try {
      const response = await fetch(fullUrl, options);
      
      // If request fails and we're not already in fallback mode, try fallback
      if (!response.ok && this.status?.current_mode !== TunnelMode.Fallback) {
        console.warn('Main domain request failed, trying fallback');
        await this.switchMode(TunnelMode.Fallback);
        
        // Retry with fallback URL
        const fallbackUrl = url.startsWith('http') || url.startsWith('tauri://') ? url : 
          (typeof window !== 'undefined' && window.__TAURI__ ? `tauri://localhost${url}` : `http://localhost:${port}${url}`);
        return await fetch(fallbackUrl, options);
      }
      
      return response;
    } catch (error) {
      // If request fails and we're not already in fallback mode, try fallback
      if (this.status?.current_mode !== TunnelMode.Fallback) {
        console.warn('Main domain request failed, trying fallback');
        await this.switchMode(TunnelMode.Fallback);
        
        const fallbackUrl = url.startsWith('http') || url.startsWith('tauri://') ? url : 
          (typeof window !== 'undefined' && window.__TAURI__ ? `tauri://localhost${url}` : `http://localhost:${port}${url}`);
        return await fetch(fallbackUrl, options);
      }
      
      throw error;
    }
  }

  // Get current status for UI display
  getCurrentStatus(): TunnelStatus | null {
    return this.status;
  }

  // Get current config for UI display
  getCurrentConfig(): TunnelConfig | null {
    return this.config;
  }

  // Check if tunnel is healthy
  isHealthy(): boolean {
    return this.status?.is_active ?? false;
  }

  // Get current mode for UI
  getCurrentMode(): TunnelMode {
    return this.status?.current_mode as TunnelMode ?? TunnelMode.MainDomain;
  }

  // Get response time for UI
  getResponseTime(): number {
    return this.status?.response_time_ms ?? 0;
  }

  stopMonitoring(): void {
    if (this.healthCheckInterval) {
      clearInterval(this.healthCheckInterval);
      this.healthCheckInterval = null;
    }
  }
}

export const tunnelService = new TunnelService();

File: src-tauri/src/tunnel.rs
use serde::{Deserialize, Serialize};
use std::time::{Duration, Instant};
use tokio::sync::RwLock;
use reqwest::Client;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TunnelConfig {
    pub main_domain: String,
    pub subdomain: String,
    pub target_port: u16,
    pub health_check_path: String,
    pub fallback_urls: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct TunnelStatus {
    pub is_active: bool,
    pub current_mode: TunnelMode,
    pub main_domain_status: bool,
    pub fallback_status: bool,
    pub last_check: u64,
    pub response_time_ms: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum TunnelMode {
    MainDomain,
    Fallback,
    Direct,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum TunnelError {
    ConnectionFailed,
    Timeout,
    InvalidConfig,
    AllEndpointsDown,
}

impl std::fmt::Display for TunnelError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            TunnelError::ConnectionFailed => write!(f, "Connection failed"),
            TunnelError::Timeout => write!(f, "Request timeout"),
            TunnelError::InvalidConfig => write!(f, "Invalid configuration"),
            TunnelError::AllEndpointsDown => write!(f, "All endpoints are down"),
        }
    }
}

pub struct TunnelState {
    pub config: RwLock<TunnelConfig>,
    pub status: RwLock<TunnelStatus>,
    pub client: Client,
    #[allow(dead_code)]
    pub health_check_interval: Duration,
}

impl Default for TunnelState {
    fn default() -> Self {
        let config = TunnelConfig {
            main_domain: "https://www.zentaira.com".to_string(), // www yönlendirmesi
            subdomain: "whisper-weather".to_string(),
            target_port: 3000,
            health_check_path: "/".to_string(), // Ana sayfayı kontrol et
            fallback_urls: vec![
                "https://zentaira.com".to_string(), // Fallback domain
            ],
        };

        let status = TunnelStatus {
            is_active: true,
            current_mode: TunnelMode::MainDomain,
            main_domain_status: false,
            fallback_status: false,
            last_check: 0,
            response_time_ms: 0,
        };

        Self {
            config: RwLock::new(config),
            status: RwLock::new(status),
            client: Client::new(),
            health_check_interval: Duration::from_secs(30),
        }
    }
}

#[tauri::command]
pub async fn check_tunnel_health(
    state: tauri::State<'_, TunnelState>,
) -> Result<TunnelStatus, TunnelError> {
    let config = state.config.read().await;
    let mut status = state.status.write().await;
    
    let start_time = Instant::now();
    
    // Check main domain (zentaira.com)
    let main_domain_url = format!("{}{}", config.main_domain, config.health_check_path);
    let main_domain_status = check_endpoint_health(&state.client, &main_domain_url).await;
    
    // Check fallback URLs
    let mut fallback_status = false;
    for fallback_url in &config.fallback_urls {
        let fallback_check_url = format!("{}{}", fallback_url, config.health_check_path);
        if check_endpoint_health(&state.client, &fallback_check_url).await {
            fallback_status = true;
            break;
        }
    }
    
    let response_time = start_time.elapsed().as_millis() as u64;
    
    // Update status
    let new_status = TunnelStatus {
        is_active: main_domain_status || fallback_status,
        current_mode: if main_domain_status {
            TunnelMode::MainDomain
        } else if fallback_status {
            TunnelMode::Fallback
        } else {
            TunnelMode::Direct
        },
        main_domain_status,
        fallback_status,
        last_check: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs(),
        response_time_ms: response_time,
    };
    
    let return_status = TunnelStatus {
        is_active: new_status.is_active,
        current_mode: new_status.current_mode.clone(),
        main_domain_status: new_status.main_domain_status,
        fallback_status: new_status.fallback_status,
        last_check: new_status.last_check,
        response_time_ms: new_status.response_time_ms,
    };
    
    *status = new_status;
    Ok(return_status)
}

#[tauri::command]
pub async fn get_tunnel_config(
    state: tauri::State<'_, TunnelState>,
) -> Result<TunnelConfig, TunnelError> {
    let config = state.config.read().await;
    Ok(config.clone())
}

#[tauri::command]
pub async fn update_tunnel_config(
    config: TunnelConfig,
    state: tauri::State<'_, TunnelState>,
) -> Result<(), TunnelError> {
    let mut config_state = state.config.write().await;
    *config_state = config;
    Ok(())
}

#[tauri::command]
pub async fn get_tunnel_status(
    state: tauri::State<'_, TunnelState>,
) -> Result<TunnelStatus, TunnelError> {
    let status = state.status.read().await;
    let return_status = TunnelStatus {
        is_active: status.is_active,
        current_mode: status.current_mode.clone(),
        main_domain_status: status.main_domain_status,
        fallback_status: status.fallback_status,
        last_check: status.last_check,
        response_time_ms: status.response_time_ms,
    };
    Ok(return_status)
}

#[tauri::command]
pub async fn switch_tunnel_mode(
    mode: TunnelMode,
    state: tauri::State<'_, TunnelState>,
) -> Result<(), TunnelError> {
    let mut status = state.status.write().await;
    status.current_mode = mode.clone();
    
    match mode {
        TunnelMode::MainDomain => {
            status.is_active = status.main_domain_status;
        }
        TunnelMode::Fallback => {
            status.is_active = status.fallback_status;
        }
        TunnelMode::Direct => {
            status.is_active = true;
        }
    }
    
    Ok(())
}

async fn check_endpoint_health(client: &Client, url: &str) -> bool {
    match client.get(url)
        .timeout(Duration::from_secs(3)) // 5'ten 3'e düşür
        .send()
        .await {
            Ok(response) => response.status().is_success(),
            Err(_) => false,
        }
}

// Get the best available URL for API requests
#[allow(dead_code)]
pub async fn get_best_api_url(state: &TunnelState) -> String {
    let config = state.config.read().await;
    let status = state.status.read().await;
    
    match status.current_mode {
        TunnelMode::MainDomain => format!("{}{}", config.main_domain, "/api"),
        TunnelMode::Fallback => {
            if let Some(fallback) = config.fallback_urls.first() {
                format!("{}{}", fallback, "/api")
            } else {
                format!("{}{}", config.main_domain, "/api")
            }
        }
        TunnelMode::Direct => format!("http://localhost:{}", config.target_port),
    }
}
